content/blog/2019-09-25-vril-vector-bril.md

```diff
@@ -0,0 +1,139 @@
+
+### Evaluating Arrays
+We take three approaches to evaluate arrays,
+- evaluating integration to Bril
+- qualitative evaluation to justify arrays
+- evaluating functionality of arrays
+
+#### Integration tests for arrays
+We ran tests to generate JSON representation of Bril from text format and vice-versa to evaluate whether the array specification is expressible in both available formats. These tests reside in `vtest/parse` and `vtest/print` directories.
+
+#### Qualitative evaluation of arrays
+Arrays provide a convenient method to use loops in Bril. To add two groups of 10 numbers, conventional Brili required 44 lines of code, 20 of which would define the 20 numbers and 10 to add them. Vril required 50 lines of code, 20 of which were still used to define two arrays element by element and 6 lines to add them in a loop body. It is observed that as the size of the arrays grow, conventional Bril would need correspondingly more lines of code where as Vril would still use 6 lines of code in the loop body.
+
+#### Functional evaluation of arrays
+We first wrote a simple test to initialize an array, store a value to an array and read a value from an array. 
+A second test described a loop which would add two array elements and save it in a third array.
+Both these tests generated the expected result from the interpreter. We also walked through the program to count the dynamic instruction count and number of hops over basic blocks. A program with 50 lines of code created 109 dynamic instructions and 11 hops, which matched the expectation. These tests reside in `vtest/interp` directory.
+
+### Evaluating Vector operation `vadd`
+We took a similar approach to evaluate our addition of `vadd` in Vril.
+- evaluating integration to Bril
+- evaluating functionality of `vadd`
+- evaluating performance of `vadd`
+
+#### Integration tests for Vectors
+We added tests to generate JSON and text representation of Bril from each other to gain confidence about its translatability. These tests are included in `vtest/parse` and `vtest/print`.
+
+#### Functional testing of `vadd`
+We wrote a program with `vadd` and ran this test in `vrili` with different values for `maxvl` which represents the size of vector lenght of the backend. The test generated expected results. This is included in `vtest/interp/array_vector.bril`.
+
+#### Performance of `vadd`
+Initial motivation for Vril is to reduce hop counts by having vector operators in Bril. This aspect is tested by comparing dynamic instruction count, hop count and lines of code for a loop executed in a scalar manner and a vector operation. 
+
+| Program           | Lines of code | Dynamic instructions | Hop count |
+|-------------------|---------------|----------------------|-----------|
+| Conventional Bril | 43            | 41                   | 1         |
+| Scalar Array      | 50            | 109                  | 12        |
```

Comment:
I'm somewhat surprised that the array with a loop required _more_ instructions than the version with 10 separate variables. Can you expand on why this is the case?

Reply:
Yes, this is because array programs have the overhead of initializing the arrays element-wise. I have added this description [here](https://github.com/sampsyo/cs6120/blob/ed80a1e742562fab205319bedf65c46cabbc72d0/content/blog/2019-09-25-vril-vector-bril.md#qualitative-evaluation-of-arrays), but from the [comment](https://github.com/sampsyo/cs6120/pull/45/files#r330823062) this description should be moved here or linked with this table.